### PR TITLE
Convert to controller runtime runnables

### PR DESCRIPTION
**Pull request by sunjayBhatia**
- controller runtime manages controllers and other runnables (previously
goroutines added to the server workgroup)
- leader election now uses a combination of configmap and lease objects,
in a future release we can move to just leases
- controllers and other components are currently not dependent on leader election
- other components (status writers) are set to be dependent on leader
election
- adds to the leaderelection e2e tests to check on leaderelection
configmap in addition to new lease resource and events

### CODE REVIEW

**Comment by sunjayBhatia** [comment]
overrides https://github.com/projectcontour/contour/pull/4157

File: internal/controller/controller_test.go
Comment (sunjayBhatia) [inline internal/controller/controller_test.go:49]: curious what people think about these tests, they ensure the controller is registered and that it does not require leader election, using a generated mock manager (see mock package and `go generate` command above)

File: internal/contour/handler_test.go
Comment (sunjayBhatia) [inline internal/contour/handler_test.go:22]: curious what people think of these tests for the various components, settled on this after spiking through another option which i wonder what people think of

- new object as thin wrapper around controller-runtime manager
- mainly for testing, can test by mocking controller-runtime manager (see controller tests for example)
- instead of `manager.Add(Runnable)` method, have something like `manager.Add(Runnable, requireLeaderElection bool)`

decided against this for a few reasons, mainly bc it ended up being a v thin wrapper (need to expose a lot of things from the base controller-runtime manager) and puts more logic in the `main` package, rather than closer to the components themselves which should have the knowledge about whether they need leader election or not

File: examples/contour/02-role-contour.yaml
Comment (skriss) [inline examples/contour/02-role-contour.yaml:16]: Is this for the controller-runtime manager to write events?

File: hack/actions/install-kubernetes-toolchain.sh
Comment (skriss) [inline hack/actions/install-kubernetes-toolchain.sh:61]: I believe this can be moved into `tools.go` and managed as a normal go module dependency

File: internal/controller/controller.go
Comment (skriss) [inline internal/controller/controller.go:19]: worth some comments, either here or where it's used, on why we have this.

File: internal/debug/debug.go
Comment (skriss) [inline internal/debug/debug.go:38]: nit, this comment is now out-of-date.

File: internal/httpsvc/http.go
Comment (skriss) [inline internal/httpsvc/http.go:50]: nit, comment out of date.

File: cmd/contour/serve.go
Comment (skriss) [inline cmd/contour/serve.go:172]: looks like we could almost drop the `workgroup` package entirely, but it's still used in the featuretests framework. We could do the work to fully drop it in a separate PR.

File: cmd/contour/serve.go
Comment (skriss) [inline cmd/contour/serve.go:412]: this code used to trigger a DAG rebuild immediately when a Contour instance was elected leader, which helped avoid stale status issues - I'm not seeing anything that will do that now, but could be missing it.

File: internal/featuretests/v3/featuretests.go
Comment (skriss) [inline internal/featuretests/v3/featuretests.go:158]: this is the workgroup that would have to be dropped if we wanted to drop the entire workgroup package.

**Review by skriss — CHANGES_REQUESTED** [review]
Handful of comments, but overall this is looking nice!

File: examples/contour/02-role-contour.yaml
Comment (sunjayBhatia) [reply examples/contour/02-role-contour.yaml:16]: yep, without it, leader election still works but theres errors in the logs from failures to write events

File: cmd/contour/serve.go
Comment (sunjayBhatia) [reply cmd/contour/serve.go:412]: yeah thats a good point, I'll try to think up a test that exercises that

File: examples/contour/02-role-contour.yaml
Comment (stevesloka) [reply examples/contour/02-role-contour.yaml:16]: Might be good to add this to a role instead of a cluster role, I think for this specifically, it doesn't need cluster-wide access.

File: cmd/contour/serve.go
Comment (sunjayBhatia) [reply cmd/contour/serve.go:172]: yeah, didnt drop it entirely in case we needed it for some reason, could just use some simple coordination mechanisms to run components in tests w/o the need for the workgroup code

File: internal/controller/controller.go
Comment (stevesloka) [inline internal/controller/controller.go:24]: Does this just fulfill an interface or something?

Comment (sunjayBhatia) [reply internal/controller/controller.go:24]: yep, see https://github.com/kubernetes-sigs/controller-runtime/blob/7e7bf8c1e1df991b957dfaa8b36d975740f309ab/pkg/manager/manager.go#L301-L306

could do something to ensure the type is assignable to a particular interface to do a compilation-time check, but also would need to check the return value of this method since it matters as well, which is a little unfortunate how they implemented it

File: examples/contour/02-role-contour.yaml
Comment (sunjayBhatia) [reply examples/contour/02-role-contour.yaml:16]: hm yeah thats a good point, might open a separate PR to address this since all of the RBAC for leaderelection resources could be namespace scoped, but it makes the role generation a little different

Comment (sunjayBhatia) [reply examples/contour/02-role-contour.yaml:16]: heres what things might look like if we switched to more scoped resources for the leaderelection resource RBAC: https://github.com/projectcontour/contour/pull/4204

Comment (stevesloka) [reply examples/contour/02-role-contour.yaml:16]: Yup I think that makes sense, the less permissions we need to have the better overall.

**Review by stevesloka — APPROVED** [review]
/lgtm with @skriss's comments. Nice work @sunjayBhatia!

File: internal/featuretests/v3/featuretests.go
Comment (skriss) [reply internal/featuretests/v3/featuretests.go:158]: https://github.com/projectcontour/contour/issues/4206

File: cmd/contour/serve.go
Comment (skriss) [reply cmd/contour/serve.go:172]: https://github.com/projectcontour/contour/issues/4206

File: internal/controller/gateway.go
Comment (sunjayBhatia) [inline internal/controller/gateway.go:71]: also note that these sort of watches *should* work with dynamic updates to contour configuration since we pass in a predicate func that has access to the reconciler's data, we should be able to update that data with correct concurrency/locking mechanisms (could be other solutions, this first springs to mind, theoretically should be cheap since we shouldnt get too many config crd updates)

Comment (sunjayBhatia) [reply internal/controller/gateway.go:71]: a point in favor of doubling down on controller-runtime managed controllers/runnables, in case that was a concern @projectcontour/maintainers

File: cmd/contour/serve.go
Comment (sunjayBhatia) [reply cmd/contour/serve.go:412]: the e2e test i can think of is a little racy:
1. delete leader pod
2. make a change to an httpproxy/gateway/gatewayclass/etc. that would make it invalid
3. eventually resource should have correct status

however, 2. needs to be timed just right, after the original leader has stopped but before a new leader has been elected

Comment (sunjayBhatia) [reply cmd/contour/serve.go:412]: simple fix would be to replicate what we had before, and pass in the manager "elected" channel to the event handler, so it re-runs on leader election

longer-term, seems we might need to have coverage over this logic with a controller-runtime style test of the controller itself: https://github.com/projectcontour/contour/blob/7bf2309d411a11a80b1e0f75918f9ea5b8932d78/internal/controller/gateway.go#L89-L107 (and propagate it to other controllers, not just gateway and gatewayclass)

this way we can more easily control and test how these processes are coordinated

we might also want to possibly move other things to controllers rather than informer-based, e.g. httpproxy/ingress/secrets etc. to standardize testing

Comment (skriss) [reply cmd/contour/serve.go:412]: One idea for the E2E, could you manually create the leader-elect configmap before starting up Contour, indicating some non-existent pod holds the lease? That way no Contour pod would acquire it for at least a few seconds. Still some raciness, but maybe would be easier than having to delete a pod. Food for thought.

Comment (sunjayBhatia) [reply cmd/contour/serve.go:412]: another race i think is that even if you do trigger a dag update/status update handler update on leader election, if that isn't running yet (bc some processes start later), you might miss a write to appropriate channels and miss an update

**Comment by skriss** [comment]
@sunjayBhatia I was playing with this locally and had a couple issues:
- proxies created *after* my local Contour acquires leadership do not seem to trigger a DAG rebuild/get status
- the process is not shutting down cleanly on SIGTERM

I imagine the first is related to the E2E failures. I haven't had a chance to dig into these any further yet, just wanted to put the reports here. I'll look some more later this AM.

**Comment by sunjayBhatia** [comment]
> @sunjayBhatia I was playing with this locally and had a couple issues:
> 
> * proxies created _after_ my local Contour acquires leadership do not seem to trigger a DAG rebuild/get status
> * the process is not shutting down cleanly on SIGTERM
> 
> I imagine the first is related to the E2E failures. I haven't had a chance to dig into these any further yet, just wanted to put the reports here. I'll look some more later this AM.

yep, same issues here, e2e tests do some capturing of the issues but a little worried about other components besides just the main status updater (e.g load balancer status updater), needs a little more work

File: cmd/contour/serve.go
Comment (skriss) [reply cmd/contour/serve.go:412]: Another idea for the implementation here would be to bring back the event handler's `UpdateNow` function, then define a new leader-elected Runnable that, when started, just calls `eventHandler.UpdateNow()`. Shouldn't be much different from what you have now, but maybe a little simpler?

**Comment by sunjayBhatia** [comment]
> `Contour server failed" error="failed waiting for all runnables to end within grace period of 30s: context deadline exceeded` on shutdown, could adjust to not use graceful shutdown as well

i think fixing my silly channels mistake also helped with shutdown issues

File: internal/k8s/status_test.go
Comment (skriss) [inline internal/k8s/status_test.go:22]: It might be good to also verify in these tests that e.g. `StatusUpdateHandler` implements the `LeaderElectionRunnable` interface, just to ensure no silly typos or anything

**Review by skriss — COMMENTED** [review]
One thought on the tests, otherwise I think this looks good, doing some ad-hoc testing locally  too

File: cmd/contour/serve.go
Comment (sunjayBhatia) [reply cmd/contour/serve.go:412]: https://github.com/projectcontour/contour/pull/4202/commits/602f8f7c1d1258136d3d7c4ff2f686c7865513c0

wdyt, used that idea ^ rather than having channels flying around

File: internal/leadership/notifier.go
Comment (sunjayBhatia) [inline internal/leadership/notifier.go:35]: ran these non-blocking, dont think this should cause any issues, but worth thinking about 🤔

Comment (skriss) [reply internal/leadership/notifier.go:35]: I can't think of any issues with it 👍

**Review by skriss — APPROVED** [review]
:+1: local testing all looks good to me, as does the new leader election notification abstraction

**Comment by sunjayBhatia** [comment]
Tests need a small fix to make them more robust now we’re deleting contour and restarting in incluster suite

**Review by skriss — APPROVED** [review]
LGTM, not sure if there were any other outstanding threads here though

**Comment by sunjayBhatia** [comment]
> LGTM, not sure if there were any other outstanding threads here though

i think we got everything!

**Pull request merged**
